### PR TITLE
Add the option to minimize structures with coordgen

### PR DESCRIPTION
**Pull request by greglandrum**
This is for cases where we don't want completely new 2D coords but are interested in just cleaning up clashes and things coordgen finds "ugly".

The result isn't perfect since coordgen will still change ring systems that use standard templates it doesn't recognize (like norborane), but it's still nice to be able to do this.

### CODE REVIEW

**Comment by greglandrum** [comment]
@d-b-w : would you mind reviewing this one since it's connected to coordgen?
@ZontaNicola: FYI

**Comment by greglandrum** [comment]
The CI failure with the windows DLL build will be fixed by this coordgen PR:
https://github.com/schrodinger/coordgenlibs/pull/76
It shouldn't prevent review/merging

File: External/CoordGen/CoordGen.h [hunk 1]
```diff
@@ -156,19 +210,30 @@ unsigned int addCoords(T& mol, const CoordGenParams* params = nullptr) {
...
   for (size_t i = 0; i < mol.getNumAtoms(); ++i) {
-    conf->setAtomPos(
-        i, RDGeom::Point3D(ats[i]->coordinates.x() / scaleFactor,
-                           ats[i]->coordinates.y() / scaleFactor, 0.0));
+    auto coords = RDGeom::Point3D(ats[i]->coordinates.x() / scaleFactor,
+                                  ats[i]->coordinates.y() / scaleFactor, 0.0);
+    if (params->minimizeOnly) {
+      // readjust the average bond length from 1 (coordgen's target)
+      // to whatever we started with
+      coords *= avgBondLen;
```
Comment (bp-kelley) [inline External/CoordGen/CoordGen.h:226]: avgBondLen could be shoved into scaleFactor.

File: External/CoordGen/CoordGen.h [hunk 2]
```diff
@@ -70,24 +78,42 @@ unsigned int addCoords(T& mol, const CoordGenParams* params = nullptr) {
...
       hasTemplateMatch = true;
     }
   }
 
+  // if we're doing coordinate minimization it makes our life easier to
+  // start by translating to the origin
+  RDGeom::Point3D centroid{0.0, 0.0, 0.0};
+  if (params->minimizeOnly) {
+    auto conf = mol.getConformer();
+    for (unsigned int i = 0; i < mol.getNumAtoms(); ++i) {
```
[the remark this answers is not in the document]
Comment (bp-kelley) [reply External/CoordGen/CoordGen.h:93]: I like this as well.

File: External/CoordGen/CoordGen.h [hunk 3]
```diff
@@ -132,6 +172,20 @@ unsigned int addCoords(T& mol, const CoordGenParams* params = nullptr) {
...
+    if (params->minimizeOnly) {
+      bondLenAccum += (mol.getConformer().getAtomPos(obnd->getBeginAtomIdx()) -
+                       mol.getConformer().getAtomPos(obnd->getEndAtomIdx()))
+                          .length();
+    }
+  }
+  double avgBondLen = 1.0;
+  if (params->minimizeOnly) {
+    if (nSingle) {
+      avgBondLen = singleBondLenAccum / nSingle;
```
Comment (bp-kelley) [inline External/CoordGen/CoordGen.h:184]: Will this value every be used in practice?  I'm assuming mol.getNumBonds() is always > 1 and will override this value below.

File: External/CoordGen/CoordGen.h [hunk 4]
```diff
@@ -132,6 +172,20 @@ unsigned int addCoords(T& mol, const CoordGenParams* params = nullptr) {
...
+    }
+    if (mol.getNumBonds()) {
```
Comment (bp-kelley) [inline External/CoordGen/CoordGen.h:186]: Should this. be else if?

File: External/CoordGen/CoordGen.h (hunk 1 above)
Comment (greglandrum) [reply External/CoordGen/CoordGen.h:226]: indeed, but I think having it explicit is useful

File: External/CoordGen/CoordGen.h (hunk 3 above)
Comment (greglandrum) [reply External/CoordGen/CoordGen.h:184]: right, there is a missing "else" here

File: External/CoordGen/CoordGen.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (greglandrum) [reply External/CoordGen/CoordGen.h]: good point, that's now done (this was a few days ago, but I forgot to comment)

**Pull request merged**